rpc/src/optimistically_confirmed_bank_tracker.rs

```diff
@@ -101,31 +104,74 @@ impl OptimisticallyConfirmedBankTracker {
+            } else if bank.slot() > bank_forks.read().unwrap().root_bank().slot() {
+                pending_optimistically_confirmed_banks.insert(bank.slot());
+                debug!("notify_or_defer defer notifying for slot {:?}", bank.slot());
+            }
+        }
+    }
+
     pub fn process_notification(
         notification: BankNotification,
         bank_forks: &Arc<RwLock<BankForks>>,
         optimistically_confirmed_bank: &Arc<RwLock<OptimisticallyConfirmedBank>>,
         subscriptions: &Arc<RpcSubscriptions>,
-        pending_optimistically_confirmed_banks: &mut HashSet<Slot>,
+        mut pending_optimistically_confirmed_banks: &mut HashSet<Slot>,
+        highest_confirmed_slot: &mut Slot,
     ) {
         debug!("received bank notification: {:?}", notification);
         match notification {
             BankNotification::OptimisticallyConfirmed(slot) => {
-                if let Some(bank) = bank_forks
-                    .read()
-                    .unwrap()
-                    .get(slot)
-                    .filter(|b| b.is_frozen())
-                {
+                if let Some(bank) = bank_forks.read().unwrap().get(slot) {
                     let mut w_optimistically_confirmed_bank =
                         optimistically_confirmed_bank.write().unwrap();
-                    if bank.slot() > w_optimistically_confirmed_bank.bank.slot() {
+
+                    if bank.slot() > w_optimistically_confirmed_bank.bank.slot() && bank.is_frozen()
+                    {
                         w_optimistically_confirmed_bank.bank = bank.clone();
-                        subscriptions.notify_gossip_subscribers(slot);
+                    }
+
+                    if slot > *highest_confirmed_slot {
+                        for parent in bank.parents().iter().rev() {
+                            if parent.slot() > *highest_confirmed_slot {
+                                debug!("Calling notify_or_defer for parent {:?}", parent.slot());
+                                Self::notify_or_defer(
+                                    subscriptions,
+                                    bank_forks,
+                                    parent,
+                                    *highest_confirmed_slot,
+                                    &mut pending_optimistically_confirmed_banks,
+                                );
+                            }
+                        }
+                        Self::notify_or_defer(
+                            subscriptions,
+                            bank_forks,
+                            bank,
+                            *highest_confirmed_slot,
+                            &mut pending_optimistically_confirmed_banks,
+                        );
+
+                        *highest_confirmed_slot = slot;
                     }
                     drop(w_optimistically_confirmed_bank);
                 } else if slot > bank_forks.read().unwrap().root_bank().slot() {
```

Comment:
Hmm, I think this solution doesn't handle this case properly.
Imagine BankForks contains slots 0-2, root 0. The first optimistic confirmation we receive is for slot 3, but it's not actually in BankForks yet. We hit this case and add slot 3 to the pending list. However, slots 1-2 are not in the pending list, so when slot 3 is frozen, only slot 3 is notified.

I think that we need to use `highest_confirmed_slot` to notify parents in `BankNotification::Frozen(bank)` case below, and only update it on `notify` (not `defer`).
Then, rather than adding all the parents to the pending list in `defer` above, just add `bank.slot()`.

Reply:
You could likely demonstrate this failing in your test by not inserting bank3 into bank_forks until somewhere around here: https://github.com/solana-labs/solana/pull/19092/files#diff-836b3beddc73d5a3dbe035ef96380aab3ac7caadd149ab405b1c2534dd90b2f3R1775

Reply:
Good point, in the tests I have done against testnet. I was not able to see such conditions. @carllin Your thought? Can we have a OptimisticallyConfirmed notification without having the slot in the BankForks?

Reply:
Great catch @CriesofCarrots! @lijunwangs yes, as @CriesofCarrots has pointed out, it is possible to have a notification without having the slot in bank forks if the node is processing slightly behind the rest of the cluster

Reply:
Okay. I will look into how to handle this case.

Reply:
>  Can we have a OptimisticallyConfirmed notification without having the slot in the BankForks

@lijunwangs yeah, this can happen if we get the notification before we start replaying hte slot

Reply:
Thanks for catching and confirming this use case @CriesofCarrots, @jstarry and @carllin -- it will be handled as well.

Reply:
This is addressed in the latest update